src/distributors/adoptopenjdk-installer.ts

```diff
@@ -0,0 +1,83 @@
+import * as core from '@actions/core';
+import * as tc from '@actions/tool-cache';
+
+import fs from 'fs';
+import path from 'path';
+import semver from 'semver';
+
+import { IS_WINDOWS, PLATFORM, IS_MACOS, macOSJavaContentDir } from "../util";
+import { IJavaInfo, IJavaRelease, JavaBase, JavaInitOptions } from "./base-installer";
+import { IRelease, IReleaseVersion } from './adoptopenjdk-models'
+
+export class AdoptOpenJdkDistributor extends JavaBase {
+    private platform: string;
+    
+    constructor(initOptions: JavaInitOptions) {
+        super(initOptions);
+        this.platform = IS_MACOS ? 'mac' : PLATFORM;
+    }
+
+    private async getAvailableMajor(range: semver.Range) {
+        const urlReleaseVersion = "https://api.adoptopenjdk.net/v3/info/available_releases"
+        const javaVersionAvailable = (await this.http.getJson<IReleaseVersion>(urlReleaseVersion)).result;
+
+        if (!javaVersionAvailable) {
+            throw new Error(`No versions were found for ${this.Distributor}`)
+        }
+
+        const javaSemVer = javaVersionAvailable.available_releases.map(item => semver.coerce(item)!)!;
+        const majorVersion = semver.maxSatisfying(javaSemVer, range)?.major;
+
+        if(!majorVersion) {
+            throw new Error(`Could find version which satisfying. Versions: ${javaVersionAvailable.available_releases}`);
+        }
+
+        return majorVersion;
+    }
+
+    protected async downloadTool(javaRelease: IJavaRelease): Promise<IJavaInfo> {
+        let toolPath: string;
+        let downloadDir: string;
+
+        core.info(`Downloading ${this.Distributor}, java version ${javaRelease.resolvedVersion}`);
+        const javaPath = await tc.downloadTool(javaRelease.link);
+        
+        if(IS_WINDOWS) {
+            downloadDir = await tc.extractZip(javaPath);
+        } else {
+            downloadDir = await tc.extractTar(javaPath);
+        }
+
+        const archiveName = fs.readdirSync(downloadDir)[0];
+        const archivePath = path.join(downloadDir, archiveName);
+        toolPath = await tc.cacheDir(archivePath, `Java_${this.Distributor}_${this.javaPackage}`, javaRelease.resolvedVersion, this.arch);
+
+        if(process.platform === 'darwin') {
+            toolPath = path.join(toolPath, macOSJavaContentDir);
+        }
+
+        return { javaPath: toolPath, javaVersion: javaRelease.resolvedVersion };
+    }
+
+    protected get Distributor(): string {
```

Comment:
I think it was better when we just pass distributor name as string in contructor as you have implemented initially. Sorry for confusing. Let's use your approach

Reply:
Done.